axon-cosmosdb/src/main/java/org/axonframework/extensions/cosmosdb/eventsourcing/tokenstore/CosmosTokenStore.java

```diff
@@ -0,0 +1,516 @@
+     * <p>
+     * Will assert that the {@link CosmosAsyncClient} and {@link Serializer} are not {@code null}, and will throw an
+     * {@link AxonConfigurationException} if any of them is {@code null}.
+     *
+     * @param builder the {@link Builder} used to instantiate a {@link CosmosTokenStore} instance
+     */
+    protected CosmosTokenStore(Builder builder) {
+        builder.validate();
+        this.client = builder.client;
+        this.serializer = builder.serializer;
+        this.claimTimeout = builder.claimTimeout;
+        this.nodeId = builder.nodeId;
+        this.database = createDatabaseIfNotExists(builder.databaseName);
+    }
+
+    /**
+     * Instantiate a Builder to be able to create a {@link CosmosTokenStore}.
+     * <p>
+     * The {@code claimTimeout} is defaulted to a 10 seconds duration (by using {@link Duration#ofSeconds(long)},
+     * {@code nodeId} is defaulted to the {@code ManagementFactory#getRuntimeMXBean#getName} output, the {@link String}
+     * database name defaults to 'axon'. The {@link CosmosAsyncClient} and {@link Serializer} are <b>hard
+     * requirements</b> and as such should be provided.
+     *
+     * @return a Builder to be able to crete a {@link CosmosTokenStore}
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public void initializeTokenSegments(@Nonnull String processorName, int segmentCount)
+            throws UnableToClaimTokenException {
+        initializeTokenSegments(processorName, segmentCount, null);
+    }
+
+    @Override
+    public void initializeTokenSegments(@Nonnull String processorName,
+                                        int segmentCount,
+                                        @Nullable TrackingToken initialToken) throws UnableToClaimTokenException {
+        if (fetchSegments(processorName).length > 0) {
+            throw new UnableToClaimTokenException(
+                    "Unable to initialize segments. Some tokens were already present for the given processor."
+            );
+        }
+        IntStream.range(0, segmentCount).forEach(
+                i -> initializeSegment(initialToken, processorName, i)
+        );
+    }
+
+    @Override
+    public void storeToken(TrackingToken token, @Nonnull String processorName, int segment)
+            throws UnableToClaimTokenException {
+        try {
+            getCurrent(processorName, segment)
+                    .flatMap(current -> {
+                        validateCurrent(current.getItem());
+                        CosmosTokenItem newItem = createNewItem(token, segment);
+                        return getContainer(processorName).replaceItem(
+                                newItem,
+                                newItem.getId(),
+                                new PartitionKey(newItem.getId()),
+                                new CosmosItemRequestOptions().setIfMatchETag(current.getETag())
```

Comment:
I did a little digging to figure out what this is for, where the [following](https://www.annytab.com/safe-update-in-cosmos-db-with-etag-asp-net-core/) post helped most. Makes sense to do this by the way!

Given its importance and occurrences (four times), it might be nice to make a separate function out of this, with a little explanation. 

Granted, people who know CosmosDb would be familiar with this. 
However, ComsosDb is, at least at this stage, rather new to the Axon Framework team. 
As such, I think it would be helpful for current and future us if we have that description somewhere.

Reply:
True, extracted to a `replaceWithConcurrencyControl` method.